### PR TITLE
Two-particle transverse momentum correlations task first commit

### PR DESCRIPTION
**Pull request by victor-gonzalez**
Is the skeleton for the data collection.
Will provide also the Generalized Balance Functions

### CODE REVIEW

**Comment by jgrosseo** [comment]
Thanks for the commit. Please run clang-format on (all) your files (see the output of the failing travis CI test)

**Comment by victor-gonzalez** [comment]
Just my fault. I closed it to include the change after clang-formater when I realized it was already incorporated

**Review by jgrosseo — APPROVED** [review]

File: Analysis/Tasks/PWGCF/dptdptcorrelations.cxx
Comment (jgrosseo) [inline Analysis/Tasks/PWGCF/dptdptcorrelations.cxx:180]: You can do a full declaration of the histogram here, and then you don't need any line in init(). For example:
OutputObj<TH2F> etaphiH{TH2F("etaphi", "etaphi", 100, 0., 2. * M_PI, 102, -2.01, 2.01)};

File: Analysis/Tasks/PWGCF/dptdptcorrelations.cxx
Comment (jgrosseo) [inline Analysis/Tasks/PWGCF/dptdptcorrelations.cxx:326]: It doesn't work for you?

File: Analysis/Tasks/PWGCF/dptdptcorrelations.cxx
Comment (victor-gonzalez) [reply Analysis/Tasks/PWGCF/dptdptcorrelations.cxx:180]: Yes, you are right. I used that way in the QA task.
I'm trying few things I'm not sure they will work in the grid/train scenarios and this will be probably one of them.
The plan is, anyway, to use the HistogramRegistry I guess

File: Analysis/Tasks/PWGCF/dptdptcorrelations.cxx
Comment (victor-gonzalez) [reply Analysis/Tasks/PWGCF/dptdptcorrelations.cxx:326]: It fails at compile time.
I'm not sure if it is because Tracks1 and Tracks2 are different partitions on the same table or because I'm using the combination policy in a wrong way.

**Comment by jgrosseo** [comment]
@victor-gonzalez checkcode and gpu build will fail, and this is unrelated to your code at present.

**Comment by victor-gonzalez** [comment]
Incorporated PWGDQ directory in my modified CMakeLists.txt to resolve the merge conflict

File: Analysis/Tasks/PWGCF/dptdptcorrelations.cxx
Comment (aalkin) [inline Analysis/Tasks/PWGCF/dptdptcorrelations.cxx:150]: This may be better implemented with partitions - instead of checking each track you can immediately get groups for each of the cases.

File: Analysis/Tasks/PWGCF/dptdptcorrelations.cxx
Comment (aalkin) [inline Analysis/Tasks/PWGCF/dptdptcorrelations.cxx:88]: What is the reason to use these static pointers? They are not shared between the tasks, since those run in separate processes, each one will have their own.

File: Analysis/Tasks/PWGCF/dptdptcorrelations.cxx
Comment (aalkin) [inline Analysis/Tasks/PWGCF/dptdptcorrelations.cxx:117]: This can definitely be implemented as a filter.

File: Analysis/Tasks/PWGCF/dptdptcorrelations.cxx
Comment (victor-gonzalez) [reply Analysis/Tasks/PWGCF/dptdptcorrelations.cxx:150]: As it is now it is only checking track charge but the final implementation will need to consider also PID and, perhaps, other configurations. That's why I used this approach for the time being. If finally every is feasible with partitions I will definitely use it.

Comment (aalkin) [reply Analysis/Tasks/PWGCF/dptdptcorrelations.cxx:150]: I see. Then you may consider using 'std::pair' here, instead of tuple.

File: Analysis/Tasks/PWGCF/dptdptcorrelations.cxx
Comment (victor-gonzalez) [reply Analysis/Tasks/PWGCF/dptdptcorrelations.cxx:88]: Perhaps I misunderstood something here. Do you mean that the OutputObject that are defined at the task level will be shared by all instances of the same task in a parallel running scenario with each task receiving a different input data file?

File: Analysis/Tasks/PWGCF/dptdptcorrelations.cxx
Comment (victor-gonzalez) [reply Analysis/Tasks/PWGCF/dptdptcorrelations.cxx:117]: I agree

File: Analysis/Tasks/PWGCF/dptdptcorrelations.cxx
Comment (victor-gonzalez) [reply Analysis/Tasks/PWGCF/dptdptcorrelations.cxx:150]: I guess the usage is the same because I latter on use to populate two additional data columns which then I use to produce partitions on the tracks table.

Comment (victor-gonzalez) [reply Analysis/Tasks/PWGCF/dptdptcorrelations.cxx:150]: By the way, as you can see on the commented L.326, I tried to use these partitions to extract combinations out of them but it was rejected by the compiler and I implemented the traditional loop instead

File: Analysis/Tasks/PWGCF/dptdptcorrelations.cxx
Comment (aalkin) [reply Analysis/Tasks/PWGCF/dptdptcorrelations.cxx:88]: OutputObj has a smart pointer managed by framework, I see that most of the histograms you have can be statically declared directly as OutputObj, which already behave as pointers and have correct ownership. 

Each "timeframe" or input file will result in a separate OutputObj for a given task, of course, they are not directly shared and their contents is automatically merged each time task finished processing a file.  I'm just wondering why do you need these pointers separately? Inside process you can always do `auto obj = *outputobj;` if you need the contained object itself, or you can call that contained object methods with `outputobj->method();`.

File: Analysis/Tasks/PWGCF/dptdptcorrelations.cxx
Comment (aalkin) [reply Analysis/Tasks/PWGCF/dptdptcorrelations.cxx:150]: I see, I'll try to check what is the problem with those.

**Comment by aalkin** [comment]
Just as a general comment, try to avoid doing branch merging, it can create problematic commit histories and lead to merge conflicts for your final PR. The better way to keep current with the central dev branch is to use `git rebase`.

File: Analysis/Tasks/PWGCF/dptdptcorrelations.cxx
Comment (victor-gonzalez) [reply Analysis/Tasks/PWGCF/dptdptcorrelations.cxx:88]: Ok! That's what I thought.
Probably this global pointers will end up within the task scope. I have them here global because before I had two tasks one before event selection and another one after event selection which filled a different set of histograms.

But I definitely wanted them configured at task init time because the histograms will be configuration dependent. For instance, if the task is run for pp the histograms will be different as it is run for PbPb or if it is run for MC reco level or MC gen level. I want to have the histogram construction depending on the configuration, or even the running conditions, on the init section of the task and not 'hard' coded.

This also reminds me the limitation in the number of histograms which will be fixed with the HistogramRegistry.
If a TList can be used as an OutputObj we could use any number of histograms not declaring them as OutputObj just appending them to the TList as far a we are sure we are using the same list structure along all task instances. I think that was one of the things I was trying when I realized that a TList cannot be used as an OutputObj

**Comment by victor-gonzalez** [comment]
Thanks for the suggestion.
Still learning git(hub)

**Review by jgrosseo — APPROVED** [review]

**Comment by jgrosseo** [comment]
@victor-gonzalez FYI: As this is your first commit, the tests only start after approval. Once this has been merged, the tests will always start right away.

**Comment by jgrosseo** [comment]
The failures are unrelated. @aalkin Seeing your comments, are there updates you would like to wait for before I merge?

File: Analysis/Tasks/PWGCF/dptdptcorrelations.cxx
Comment (aalkin) [reply Analysis/Tasks/PWGCF/dptdptcorrelations.cxx:88]: Tlist can be used with OutputObj! There is an example in histograms.cxx.

**Comment by aalkin** [comment]
@jgrosseo it is functional, so it should be merged to be tested on a train. More development is needed, but it is mostly held back by framework issues.

**Pull request merged**
